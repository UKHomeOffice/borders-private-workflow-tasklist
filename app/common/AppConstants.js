
export default class AppConstants {
    static DASHBOARD_PATH = "/dashboard";
    static YOUR_TASKS_PATH = "/your-tasks";
    static YOUR_GROUP_TASKS_PATH = '/your-group-tasks';
    static PROCEDURES_PATH = '/procedures';
    static REPORTS_PATH = '/reports';
    static REPORT_PATH = "/report";
    static MESSAGES_PATH = "/messages";
    static CALENDAR_PATH = "/calendar";
    static START_A_PROCEDURE = "/start-a-procedure";
    static TASK_PATH="/task";
    static ADMIN_PATH="/admin";
    static SHIFT_PATH = "/shift";
    static PROCEDURE_DIAGRAM_PATH = "/procedure-diagram";
    static ONBOARD_USER_PATH="/onboard-user";
    static MANDATORY_DECLARATION_PATH="/mandatory-declarations";
    static  ONE_MINUTE = 60000;

}
