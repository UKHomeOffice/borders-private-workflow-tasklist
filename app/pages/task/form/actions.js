import * as types from './actionTypes';

const fetchTaskForm = (task) => ({
  type: types.FETCH_TASK_FORM,
  task
});


const fetchTaskFormSuccess = payload => ({
  type: types.FETCH_TASK_FORM_SUCCESS,
  payload
});

const fetchTaskFormFailure = () => ({
  type: types.FETCH_TASK_FROM_FAILURE
});


const submitTaskForm = (formId, taskId, submission, variableName) => ({
  type: types.SUBMIT_TASK_FORM,
  formId,
  taskId,
  submission,
  variableName
});

const submitTaskFormSuccess = (payload) => ({
  type: types.SUBMIT_TASK_FORM_SUCCESS,
  payload
});

const submitTaskFormFailure = () => ({
  type: types.SUBMIT_TASK_FORM_FAILURE
});

const completeTask = (taskId, data) => ({
  type: types.COMPLETE_TASK_FORM,
  taskId,
  data
});

const completeTaskSuccess = (payload) => ({
  type: types.COMPLETE_TASK_FORM_SUCCESS,
  payload
});

const completeTaskFailure = () => ({
  type: types.COMPLETE_TASK_FORM_FAILURE
});

const resetForm = () => ({
  type: types.RESET_FORM
});


const customEvent = (event, task, variableName) => ({
  type: types.TASK_CUSTOM_EVENT,
  event,
  task,
  variableName
});

const customEventSuccess = (payload) => ({
  type: types.TASK_CUSTOM_EVENT_SUCCESS,
  payload
});

const customEventFailure = (error, event) => ({
  type: types.TASK_CUSTOM_EVENT_FAILURE,
  error,
  event
});

export {
  fetchTaskForm,
  fetchTaskFormSuccess,
  fetchTaskFormFailure,
  submitTaskForm,
  submitTaskFormSuccess,
  submitTaskFormFailure,
  completeTask,
  completeTaskSuccess,
  completeTaskFailure,
  resetForm,
  customEvent,
  customEventFailure,
  customEventSuccess
};
