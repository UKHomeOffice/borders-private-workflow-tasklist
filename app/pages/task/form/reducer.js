import Immutable from 'immutable';
import * as actions from './actionTypes';
import {
  FAILED, NOT_SUBMITTED, SUBMISSION_SUCCESSFUL, SUBMITTING,
} from './constants';

const { Map } = Immutable;

const initialState = new Map({
  loadingTaskForm: false,
  form: null,
  submissionStatus: NOT_SUBMITTED,
  customEventSubmissionStatus: NOT_SUBMITTED,
  submissionResponse: null,
  nextTask: null,
  nextVariables: null
});

function reducer(state = initialState, action) {
  switch (action.type) {
    case actions.RESET_FORM:
      return initialState;
    case actions.FETCH_TASK_FORM:
      return state.set('loadingTaskForm', true)
        .set('form', null)
        .set('taskFormCompleteSuccessful', false)
        .set('submittingTaskFormForCompletion', false);
    case actions.FETCH_TASK_FORM_SUCCESS:
      const data = action.payload.entity;
      return state.set('loadingTaskForm', false)
        .set('form', data);
    case actions.FETCH_TASK_FROM_FAILURE:
      return state.set('loadingTaskForm', false);
    case actions.SUBMIT_TASK_FORM:
      return state.set('submissionStatus', SUBMITTING);
    case actions.SUBMIT_TASK_FORM_FAILURE:
      return state.set('submissionStatus', FAILED);
    case actions.COMPLETE_TASK_FORM:
      return state.set('submissionStatus', SUBMITTING);
    case actions.COMPLETE_TASK_FORM_SUCCESS:
      if (action.payload.entity !== '') {
        const rawVariables =  action.payload.entity.variables? action.payload.entity.variables : {};
        const variables = {};
        Object.keys(rawVariables).forEach(key => {
          if (rawVariables[key].type === 'Json') {
            variables[key] = JSON.parse(rawVariables[key].value);
          } else {
            variables[key] = rawVariables[key].value;
          }
        });
        action.payload.entity.variables = variables;
      }
      return state.set('submissionStatus', SUBMISSION_SUCCESSFUL)
          .set('submissionResponse', action.payload.entity);
    case actions.COMPLETE_TASK_FORM_FAILURE:
      return state.set('submissionStatus', FAILED);
    case actions.SET_NEXT_TASK:
      const {task} = action;
      return state.set('nextTask', Immutable.fromJS(task))
          .set('nextVariables', action.variables);
    case actions.TASK_CUSTOM_EVENT:
      return state.set('customEventSubmissionStatus', SUBMITTING);
    case actions.TASK_CUSTOM_EVENT_SUCCESS:
      return state.set('customEventSubmissionStatus', SUBMISSION_SUCCESSFUL);
    case actions.TASK_CUSTOM_EVENT_FAILURE:
      return state.set('customEventSubmissionStatus', FAILED);
    default:
      return state;
  }
}

export default reducer;
