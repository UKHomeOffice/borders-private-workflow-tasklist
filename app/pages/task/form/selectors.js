import { NAME } from './constants';

export const form = state => state[NAME].get('form');
export const loadingTaskForm = state => state[NAME].get('loadingTaskForm');
export const submissionStatus = state => state[NAME].get('submissionStatus');
export const customEventSubmissionStatus = state => state[NAME].get('customEventSubmissionStatus');
