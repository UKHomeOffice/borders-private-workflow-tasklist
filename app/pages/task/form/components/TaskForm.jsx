import React from 'react';
import {Form} from 'react-formio';
import AppConstants from '../../../../common/AppConstants';
import GovUKDetailsObserver from "../../../../core/util/GovUKDetailsObserver";
import FormioEventListener from "../../../../core/util/FormioEventListener";
import FileService from "../../../../core/FileService";

export default class TaskForm extends React.Component {

    constructor(props) {
        super(props);
        this.formNode = React.createRef();
    }

    componentDidMount() {
        this.observer = new GovUKDetailsObserver(this.formNode.element).create();
    }

    componentWillUnmount() {
        this.observer.destroy();
    }

    shouldComponentUpdate(nextProps, nextState, nextContext) {
        return false;
    }

    handleCancel = (resetForm) => {
        this.props.history.replace(AppConstants.DASHBOARD_PATH);
        resetForm(false);
    };


    render() {
        const {onCustomEvent, variables, task, onSubmitTaskForm, formReference, form, kc} = this.props;
        const formVariableSubmissionName = `${form.name}::submissionData`;
        const options = {
            noAlerts: true,
            hooks: {
                beforeCancel: (...args) => {
                    this.handleCancel(args);
                },
                fileService: new FileService(kc),
                beforeSubmit: (submission, next) => {
                    submission.data.meta = {
                        formVersionId: form.versionId,
                        formId: form.id,
                        title: form.title,
                        name: form.name,
                        submissionDate: new Date(),
                        submittedBy: kc.tokenParsed.email
                    };
                    next();
                }
            },
            i18n: {
                en: {
                    submit: 'Submit',
                }
            },
        };
        let submissionData = null;

        if (variables) {
            if (variables['submissionData']) {
                submissionData = variables['submissionData']
            } else if (variables[formVariableSubmissionName]) {
                submissionData = variables[formVariableSubmissionName];
            }
        }

        const variableInput = form.components.find(c => c.key === 'submitVariableName');
        const variableName = variableInput ? variableInput.defaultValue : form.name;

        if (!task.get('assignee')) {
            options.readOnly = true;
        }
        return <Form form={form} options={options}
                     ref={form => {
                         this.form = form;
                         formReference(form);
                         this.formNode = form;
                         if (this.form) {
                             this.form.createPromise.then(() => {
                                 new FormioEventListener(this.form, this.props);
                             });
                         }
                     }
                     }
                     onCustomEvent={(event) => onCustomEvent(event, variableName)}
                     submission={JSON.parse(submissionData)} onSubmit={(submission) => {
            onSubmitTaskForm(submission.data, variableName);
        }}/>;
    }

}
