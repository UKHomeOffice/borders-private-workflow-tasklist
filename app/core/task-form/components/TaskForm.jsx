import React, {PropTypes} from 'react'
import {
    form, loadingTaskForm, submittingTaskFormForCompletion,
    taskFormValidationSuccessful,
    taskFormCompleteSuccessful
} from "../selectors";
import {bindActionCreators} from "redux";
import * as actions from "../actions";

import {withRouter} from "react-router-dom";
import {connect} from "react-redux";
import {Form} from 'react-formio';
import {createStructuredSelector} from "reselect";
import Spinner from 'react-spinkit';

class TaskForm extends React.Component {

    componentDidMount() {
        this.props.fetchTaskForm(this.props.task);
    }


    componentWillReceiveProps(nextProps) {
        if (this.form && (this.form.formio && this.form.formio.data.submit)) {
            if (nextProps.taskFormCompleteSuccessful && nextProps.taskFormValidationSuccessful) {
                this.form.formio.emit("submitDone");
                this.props.history.replace("/tasks");
            } else {
                if (!nextProps.submittingTaskFormForCompletion) {
                    this.form.formio.emit("error");
                    this.form.formio.emit('change', this.form.formio.submission);
                }
            }
        }
    }

    componentWillUnmount() {
        this.form= null;
        this.props.resetForm();
    }

    renderForm() {
        const {loadingTaskForm, form, task, variables} = this.props;
        if (loadingTaskForm) {
            return <div>Loading form for {task.get('name')} </div>
        } else {
            const options = {
                noAlerts: true
            };

            if (!task.get('assignee')) {
                options.readOnly = true;
            }

            if (form) {
                const submit = form.components.find(c => c.key === 'submit');
                const submissionData = variables['submissionData'];
                if (submit) {
                    const variableInput = form.components.find(c => c.key === 'submitVariableName');
                    const variableName = variableInput.defaultValue;

                    if (submissionData) {
                        return <Form form={form} options={options} ref={(form) => this.form = form}
                                     submission={JSON.parse(submissionData)} onSubmit={(submission) => {
                            this.props.submitTaskForm(form._id, task.get('id'), submission.data, variableName);

                        }}/>
                    } else {
                        return <Form form={form} ref={(form) => this.form = form} options={options}
                                     onSubmit={(submission) => {
                                         this.props.submitTaskForm(form._id, task.get('id'), submission.data, variableName);
                                     }}/>
                    }

                } else {
                    return <Form form={form} options={
                       options
                    } ref={(form) => this.form = form}
                                 submission={JSON.parse(submissionData)}/>
                }


            } else {
                return <div/>
            }

        }
    }

    render() {
        const {submittingTaskFormForCompletion, submittingTaskFormForValidation} = this.props;
        return <div>
            {submittingTaskFormForValidation || submittingTaskFormForCompletion ?
                <div style={{display: 'flex', justifyContent: 'center', paddingTop: '5px'}}><Spinner
                    name="three-bounce" color="#005ea5"/></div> : <div/>
            }
            <div>{this.renderForm()}</div>
        </div>
    }
}

TaskForm.propTypes = {
    submitTaskForm: PropTypes.func.isRequired,
    fetchTaskForm: PropTypes.func.isRequired,
    resetForm: PropTypes.func.isRequired,
    loadingTaskForm: PropTypes.bool
};


const mapStateToProps = createStructuredSelector({
    form: form,
    loadingTaskForm: loadingTaskForm,
    taskFormValidationSuccessful: taskFormValidationSuccessful,
    submittingTaskFormForCompletion: submittingTaskFormForCompletion,
    taskFormCompleteSuccessful: taskFormCompleteSuccessful

});


const mapDispatchToProps = dispatch => bindActionCreators(actions, dispatch);


export default withRouter(connect(mapStateToProps, mapDispatchToProps)(TaskForm));