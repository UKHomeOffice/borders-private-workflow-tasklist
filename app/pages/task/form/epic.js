import { combineEpics } from 'redux-observable';
import PubSub from 'pubsub-js';
import * as types from './actionTypes';
import * as actions from './actions';
import { errorObservable } from '../../../core/error/epicUtil';
import { retry } from '../../../core/util/retry';


const createProcessVariables = (action, userEmail) => {
  const processVariables = {};
  processVariables[action.variableName] = {
    value: JSON.stringify(action.event.data),
    type: 'json',
  };
  processVariables.customEventUserId = {
    value: userEmail,
    type: 'string',
  };
  return processVariables;
};

const customEvent = (action$, store, { client }) => action$.ofType(types.TASK_CUSTOM_EVENT)
  .mergeMap(action => client({
    method: 'POST',
    headers: {
      Accept: 'application/json',
      'Content-Type': 'application/json',
      Authorization: `Bearer ${store.getState().keycloak.token}`,
    },
    path: `${store.getState().appConfig.workflowServiceUrl}/rest/camunda/message`,
    entity: {
      messageName: action.event.type,
      processInstanceId: action.task.get('processInstanceId'),
      processVariables: createProcessVariables(
        action, store.getState().keycloak.tokenParsed.email,
      ),
    },
  }).retryWhen(retry)
    .map(payload => {
      const { properties } = action.event.component;
      const message = properties && properties['success-message']
        ? properties['success-message'] : `${action.event.component.label} successfully actioned`;
      PubSub.publish('submission', {
        submission: true,
        autoDismiss: true,
        message,
      });
      return actions.customEventSuccess(payload);
    })
    .catch(error => errorObservable(actions.customEventFailure(error, action.event), error)));

const fetchTaskForm = (action$, store, { client }) => action$.ofType(types.FETCH_TASK_FORM)
  .mergeMap(action => client({
    method: 'GET',
    path: `${store.getState().appConfig.translationServiceUrl}/form/${action.task.get('formKey')}?taskId=${action.task.get('id')}&processInstanceId=${action.task.get('processInstanceId')}`,
    headers: {
      Accept: 'application/json',
      Authorization: `Bearer ${store.getState().keycloak.token}`,
    },
  })
    .retryWhen(retry)
    .map(payload => actions.fetchTaskFormSuccess(payload))
    .catch(error => errorObservable(actions.fetchTaskFormFailure(), error)));


const submitTaskForm = (action$, store, { client }) => action$.ofType(types.SUBMIT_TASK_FORM)
  .mergeMap(action => {
    const data = {
      variables: {},
    };
    data.variables[action.variableName] = {
      value: JSON.stringify(action.submission),
      type: 'Json',
      valueInfo: {},
    };
    return client({
      method: 'POST',
      path: `${store.getState().appConfig.workflowServiceUrl}/api/workflow/tasks/${action.taskId}/form/_complete`,
      entity: data,
      headers: {
        Accept: 'application/json',
        Authorization: `Bearer ${store.getState().keycloak.token}`,
        'Content-Type': 'application/json',
      },
    }).map(payload => {
      PubSub.publish('submission', {
        submission: true,
        autoDismiss: true,
        message: 'Task successfully completed',
      });
      return actions.completeTaskSuccess(payload);
    }).catch(error => errorObservable(actions.completeTaskFailure(), error))
  });




export default combineEpics(fetchTaskForm, submitTaskForm, customEvent);
