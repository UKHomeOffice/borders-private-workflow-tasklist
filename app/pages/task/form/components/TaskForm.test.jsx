import React from 'react';
import Immutable from 'immutable';
import AppConstants from '../../../../common/AppConstants';
import TaskForm from './TaskForm';
import form from './fixtures';

let props;

describe('TaskForm Component', () => {
  beforeEach(() => {
    props = {
      appConfig: {},
      form,
      formReference: jest.fn(),
      history: {
        replace: jest.fn(),
      },
      kc: {
        token: 'token',
        realm: 'cop-test',
        url: 'http://localhost',
        subject: 'd3eb0456-1b82-4c64-a987-7201d9ce9312',
        tokenParsed: {
          email: 'yesy',
          family_name: 'test',
          given_name: 'name',
          session_state: 'state',
          groups: ['/group/one', '/group/two'],
          realm_access: { roles: ['role'] },
        },
      },
      onCustomEvent: jest.fn(),
      onSubmitTaskForm: jest.fn(),
      task: Immutable.fromJS({
        assignee: null,
        name: 'testTask',
      }),
    };
  });

  it('renders without crashing', () => {
    const wrapper = shallow(<TaskForm {...props} />);
    expect(wrapper.exists()).toBe(true);
  });

  it('changes history prop when handleCancel called', () => {
    const wrapper = shallow(<TaskForm {...props} />);
    wrapper.instance().handleCancel(() => {});
    expect(props.history.replace).toHaveBeenCalledTimes(1);
    expect(props.history.replace).toHaveBeenCalledWith(
      AppConstants.DASHBOARD_PATH,
    );
  });

  it('matches snapshot', () => {
    const wrapper = shallow(<TaskForm {...props} />);
    expect(wrapper).toMatchSnapshot();
  });
});
