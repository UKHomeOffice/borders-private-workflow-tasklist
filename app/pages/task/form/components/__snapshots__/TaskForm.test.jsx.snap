// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskForm Component matches snapshot 1`] = `
<Form
  form={
    Object {
      "access": Array [
        Object {
          "roles": Array [
            "XXXX",
            "XXXX",
            "XXXXX",
          ],
          "type": "read_all",
        },
      ],
      "components": Array [
        Object {
          "autofocus": false,
          "clearOnHide": true,
          "conditional": Object {
            "eq": "",
            "show": "",
            "when": null,
          },
          "defaultValue": "",
          "hidden": false,
          "id": "id",
          "input": true,
          "inputMask": "",
          "inputType": "text",
          "key": "text",
          "label": "Text",
          "labelPosition": "top",
          "multiple": false,
          "persistent": true,
          "placeholder": "",
          "prefix": "",
          "properties": Object {},
          "protected": false,
          "spellcheck": true,
          "suffix": "",
          "tableView": true,
          "tags": Array [],
          "type": "textfield",
          "unique": false,
          "validate": Object {
            "custom": "",
            "customPrivate": false,
            "maxLength": "",
            "minLength": "",
            "pattern": "",
            "required": false,
          },
        },
        Object {
          "action": "event",
          "autofocus": false,
          "block": false,
          "conditional": Object {
            "eq": "",
            "show": "",
            "when": null,
          },
          "disableOnInvalid": false,
          "event": "cancel-operation",
          "id": "canceloperation",
          "input": true,
          "key": "canceloperation",
          "label": "Cancel operation",
          "leftIcon": "",
          "properties": Object {
            "success-message": "Operation successfully cancelled",
          },
          "rightIcon": "",
          "size": "md",
          "tableView": false,
          "tags": Array [],
          "theme": "primary",
          "type": "button",
        },
      ],
      "created": "2019-02-01T07:13:35.002Z",
      "display": "form",
      "id": "XXXXX",
      "machineName": "customButtonEvent",
      "modified": "2019-02-01T07:13:35.069Z",
      "name": "customButtonEvent",
      "owner": "XXXXX",
      "path": "custombuttonevent",
      "submissionAccess": Array [
        Object {
          "roles": Array [
            "5b0fa1b2769993003d6fba6f",
          ],
          "type": "create_own",
        },
        Object {
          "roles": Array [
            "XXXXX",
          ],
          "type": "read_own",
        },
        Object {
          "roles": Array [
            "XXXX",
          ],
          "type": "update_own",
        },
        Object {
          "roles": Array [
            "XXXXX",
          ],
          "type": "delete_own",
        },
      ],
      "tags": Array [
        "common",
      ],
      "title": "Custom Button Event",
      "type": "form",
    }
  }
  onChange={[Function]}
  onCustomEvent={[Function]}
  onSubmit={[Function]}
  options={
    Object {
      "fileService": FileService {
        "keycloak": Object {
          "realm": "cop-test",
          "subject": "d3eb0456-1b82-4c64-a987-7201d9ce9312",
          "token": "token",
          "tokenParsed": Object {
            "email": "yesy",
            "family_name": "test",
            "given_name": "name",
            "groups": Array [
              "/group/one",
              "/group/two",
            ],
            "realm_access": Object {
              "roles": Array [
                "role",
              ],
            },
            "session_state": "state",
          },
          "url": "http://localhost",
        },
      },
      "hooks": Object {
        "beforeCancel": [Function],
        "beforeSubmit": [Function],
      },
      "i18n": Object {
        "en": Object {
          "submit": "Submit",
        },
      },
      "noAlerts": true,
      "readOnly": true,
    }
  }
  submission={
    Object {
      "data": Object {
        "environmentContext": Object {
          "attachmentServiceUrl": undefined,
          "operationalDataUrl": undefined,
          "privateUiUrl": "http://localhost",
          "referenceDataUrl": undefined,
          "workflowUrl": undefined,
        },
        "extendedStaffDetailsContext": "",
        "keycloakContext": Object {
          "accessToken": "token",
          "email": "yesy",
          "familyName": "test",
          "givenName": "name",
          "groups": Array [
            "/group/one",
            "/group/two",
          ],
          "realm": "cop-test",
          "refreshToken": undefined,
          "roles": Array [
            "role",
          ],
          "sessionId": "state",
          "subject": "d3eb0456-1b82-4c64-a987-7201d9ce9312",
          "url": undefined,
        },
        "processContext": Object {},
        "shiftDetailsContext": "",
        "staffDetailsDataContext": "",
        "taskContext": Object {
          "assignee": null,
          "name": "testTask",
        },
      },
    }
  }
/>
`;
